ref/rust/src/bech32.rs

```diff
@@ -0,0 +1,190 @@
+
+use std::fmt;
+use std::error::Error;
+use std::ascii::AsciiExt;
+
+#[derive(Debug)]
+pub struct Bech32 {
+    hrp: String,
+    data: Vec<u8>
+}
+
+// Human-readable part and data part separator
+const SEP: char = '1';
+
+// Encoding character set. Maps data value -> char
+const CHARSET: [char; 32] = [
+    'q','p','z','r','y','9','x','8',
+    'g','f','2','t','v','d','w','0',
+    's','3','j','n','5','4','k','h',
+    'c','e','6','m','u','a','7','l'
+];
+
+// Reverse character set. Maps ASCII byte -> CHARSET index on [0,31]
+const CHARSET_REV: [i8; 128] = [
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    15, -1, 10, 17, 21, 20, 26, 30,  7,  5, -1, -1, -1, -1, -1, -1,
+    -1, 29, -1, 24, 13, 25,  9,  8, 23, -1, 18, 22, 31, 27, 19, -1,
+     1,  0,  3, 16, 11, 28, 12, 14,  6,  4,  2, -1, -1, -1, -1, -1,
+    -1, 29, -1, 24, 13, 25,  9,  8, 23, -1, 18, 22, 31, 27, 19, -1,
+     1,  0,  3, 16, 11, 28, 12, 14,  6,  4,  2, -1, -1, -1, -1, -1
+];
+
+// Generator coefficients
+const GEN: [u32; 5] = [0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3];
+
+
+pub fn encode(b: Bech32) -> EncodeResult {
+    Ok("".to_string())
+}
+
+pub fn decode(s: String) -> DecodeResult {
+    // Ensure overall length is within bounds
+    let len: usize = s.len();
+    if len < 8 || len > 90 {
+        return Err(CodingError::InvalidLength)
+    }
+
+    // Split at separator and check for two pieces
+    let parts: Vec<&str> = s.rsplitn(2, SEP).collect();
+    if parts.len() != 2 {
+        return Err(CodingError::InvalidChar)
+    }
+    let raw_hrp = parts[1];
+    let raw_data = parts[0];
+    if raw_hrp.len() < 1 || raw_data.len() < 6 {
+        return Err(CodingError::InvalidLength)
+    }
+
+    let mut has_lower: bool = false;
+    let mut has_upper: bool = false;
+    let mut hrp_bytes: Vec<u8> = Vec::new();
+    for b in raw_hrp.bytes() {
+        // Valid subset of ASCII
+        if b < 33 || b > 126 {
+            return Err(CodingError::InvalidChar)
+        }
+        let mut c = b;
+        // Lowercase
+        if b >= 97 && b <= 122 {
```

Comment:
Good call. This makes it so much easier to read the code than ASCII magic in there.

Added in latest commit.